ansible/library/announce_routes.py

```diff
@@ -330,44 +472,123 @@ def fib_t1_lag(topo, ptf_ip, no_default_route=False, action="announce"):
+
+
+"""
+For M0, we have 2 sets of routes that we are going to advertise
+    - 1st set routes are advertised by the upstream VMs (M1 devices)
+    - 2nd set routes are advertised by the downstream VMs (MX devices)
+
+The total number of routes are controlled by the colo_number, m0_number, mx_subnet_number, m0_subnet_number and number of MX devices from the topology file.
+We would have the following distribution:
+- M1 Routes:
+   - 1 default route, prefix: 0.0.0.0/0
+   - Subnet routes of M0 devices connected to M1 devices other than DUT, count: (colo_number*m0_number-1)*m0_subnet_number
+   - Subnet routes of MX devices connected to M0 devices connected M1 devices, count: (colo_number*m0_number-1)*mx_number*mx_subnet_number
+- MX Routes:
+   - Subunet routes of MX, count: mx_subnet_number
+"""
+
+
+def fib_m0(topo, ptf_ip, action="announce"):
     common_config = topo['configuration_properties'].get('common', {})
-    podset_number = common_config.get("podset_number", PODSET_NUMBER)
-    tor_number = common_config.get("tor_number", TOR_NUMBER)
-    tor_subnet_number = common_config.get("tor_subnet_number", TOR_SUBNET_NUMBER)
-    max_tor_subnet_number = common_config.get("max_tor_subnet_number", MAX_TOR_SUBNET_NUMBER)
-    tor_subnet_size = common_config.get("tor_subnet_size", TOR_SUBNET_SIZE)
+    colo_number = common_config.get("colo_number", COLO_NUMBER)
+    m0_number = common_config.get("m0_number", M0_NUMBER)
     nhipv4 = common_config.get("nhipv4", NHIPV4)
     nhipv6 = common_config.get("nhipv6", NHIPV6)
-    leaf_asn_start = common_config.get("leaf_asn_start", LEAF_ASN_START)
-    tor_asn_start = common_config.get("tor_asn_start", TOR_ASN_START)
+    m0_asn_start = common_config.get("m0_asn_start", M0_ASN_START)
+    m0_subnet_number = common_config.get("m0_subnet_number", M0_SUBNET_NUMBER)
+    m0_subnet_size = common_config.get("m0_subnet_size", M0_SUBNET_SIZE)
+    mx_subnet_size = common_config.get("mx_subnet_size", MX_SUBNET_SIZE)
+    mx_subnet_number = common_config.get("mx_subnet_number", MX_SUBNET_NUMBER)
+    mx_asn_start = common_config.get("mx_asn_start", MX_ASN_START)
 
     vms = topo['topology']['VMs']
     vms_config = topo['configuration']
-
+    mx_list = list(filter(lambda x: "MX" in x, vms_config.keys()))
+    mx_number = len(mx_list)
+
+    # In order to avoid overlapping the routes announced and the vlan of m0, get ip_start
+    vlan_configs = dict(
+        filter(lambda x: "default" not in x[0], topo["topology"]["DUT"]["vlan_configs"].items()))
+    vlan_prefixs = []
+    for _, vlans in vlan_configs.items():
+        for _, config in vlans.items():
+            vlan_prefixs.append(config["prefix"])
+    ip_base = get_next_net_of_nets(vlan_prefixs)
+
+    m1_routes_v4 = None
+    m1_routes_v6 = None
     for k, v in vms_config.items():
         vm_offset = vms[k]['vm_offset']
         port = IPV4_BASE_PORT + vm_offset
         port6 = IPV6_BASE_PORT + vm_offset
+        mx_index = v.get("mx_index", None)
```

Comment:
Why we need `mx_index`? Assume each m0 have `N` mx, can we assume the indices are [0, 1, ..., N - 1]